econml/validate/sensitivity_analysis.py

```diff
@@ -0,0 +1,156 @@
+            f"interval_type for sensitivity_interval must be 'theta' or 'ci'. Received {interval_type}")
+
+    r = 0
+    r_up = 1
+    r_down = 0
+    lb, ub = sensitivity_interval(theta, sigma, nu, cov,
+                                  alpha, 0, 0, 1, interval_type=interval_type)
+    if lb < 0 and ub > 0:
+        return 0
+
+    else:
+        if lb > 0:
+            target = 0
+            mult = 1
+            d = lb
+        else:
+            target = 1
+            mult = -1
+            d = ub
+
+    while abs(d) > 1e-6:
+        d = mult * sensitivity_interval(theta, sigma, nu, cov,
+                                        alpha, r, r, 1, interval_type=interval_type)[target]
+        if d > 0:
+            r_down = r
+        else:
+            r_up = r
+
+        r = (r_down + r_up) / 2
+
+    return r
+
+
+def dml_sensitivity_values(t_res, y_res):
+
+    t_res = t_res.reshape(-1, 1)
+    y_res = y_res.reshape(-1, 1)
+
+    theta = np.mean(y_res*t_res) / np.mean(t_res**2)  # Estimate the treatment effect
+    # Estimate the variance of the outcome residuals (after subtracting the treatment effect)
+    sigma2 = np.mean((y_res - theta*t_res)**2)
+    nu2 = 1/np.mean(t_res**2)  # Estimate the variance of the treatment
+
+    ls = np.concatenate([t_res**2, np.ones_like(t_res), t_res**2], axis=1)
+
+    G = np.diag(np.mean(ls, axis=0))  # G matrix, diagonal with means of ls
+    G_inv = np.linalg.inv(G)  # Inverse of G matrix, could just take reciprocals since it's diagonal
+
+    residuals = np.concatenate([y_res*t_res-theta*t_res*t_res,
+                                (y_res-theta*t_res) ** 2-sigma2, t_res**2*nu2-1], axis=1)  # Combine residuals
+    # Estimate the covariance matrix of the residuals
+    Ω = residuals.T @ residuals / len(residuals)
+    # Estimate the variance of the parameters
+    cov = G_inv @ Ω @ G_inv / len(residuals)
+
+    return {
+        "theta": theta,
+        "sigma": sigma2,
+        "nu": nu2,
+        "cov": cov
+    }
+
+
+def dr_sensitivity_values(Y, T, y_pred, t_pred):
+    y_pred = y_pred.squeeze(1) if np.ndim(y_pred) == 3 else y_pred
+
+    Y = Y.squeeze()
+    T_ohe = T[:, 1:].astype(int) # drop first column of T since it's the control
+    alpha = np.zeros_like(T_ohe)
+    T = np.argmax(T, axis=1)  # Convert dummy encoding to a factor vector
+
+    # one value of theta, sigma, nu for each non-control treatment
+    theta = np.zeros(T_ohe.shape[1])
+    sigma = np.zeros(T_ohe.shape[1])
+    nu = np.zeros(T_ohe.shape[1])
+
+    # one theta, sigma, nu covariance matrix for each non-control treatment
+    cov = np.zeros((T_ohe.shape[1], 3, 3))
+
+    # ATE, sigma^2, and nu^2
+    for i in range(T_ohe.shape[1]):
+        theta_score = y_pred[:, i+1] - y_pred[:, 0] + (Y-y_pred[:, i+1]) * (
+            T_ohe[:, i] == 1)/t_pred[:, i+1] - (Y-y_pred[:, 0]) * (np.all(T_ohe == 0, axis=1)/t_pred[:, 0])
+        # exclude rows with other treatments
+        sigma_score = (Y-np.choose(T, y_pred.T))**2
```

Comment:
Despite the comment, this doesn't seem to be setting the score to 0 for rows with treatments other than treatment i or control, which it should (a bad outcome prediction for treatment j shouldn't affect the sensitivity analysis for treatment i).